### DISCUSSION TITLE
SQLCipher legacy

### DISCUSSION THREAD

**Discussion by little-brother** [utelle/SQLite3MultipleCiphers#33] [General]
i'm a little confused. 
If I set `legacy` as `0` for `cipher=sqlcipher` then what option should/can I set? 
And what are their default values?

How is `legacy` related to [`cipher`-compatibility](https://www.zetetic.net/sqlcipher/sqlcipher-api/#cipher_compatibility)?

**Comment by utelle** [comment]
> If I set `legacy` as `0` for `cipher=sqlcipher` then what option should/can I set?
> And what are their default values?

The parameters you can set for the _SQLCipher_ cipher are described in the [cipher documentation for SQLCipher](https://utelle.github.io/SQLite3MultipleCiphers/docs/ciphers/cipher_sqlcipher/). At the bottom of the page you find a table with the default values of the parameters depending on the _SQLCipher_ version.

If the parameter `legacy` is set to `0`, then the default parameter values correspond typically to those of **_SQLCipher_ version 4**. However, this can be changed at compile time by selecting a different default version (by defining symbol `SQLCIPHER_VERSION_DEFAULT`). 

**Note**: Please keep in mind that using `legacy=0` results in database files that are **not compatible** with the original _SQLCipher_ library. If you need compatibility with the original _SQLCipher_ library, you **must** set `legacy` to the _SQLCipher_ version number you want compatibility with.

> How is `legacy` related to [`cipher`-compatibility](https://www.zetetic.net/sqlcipher/sqlcipher-api/#cipher_compatibility)?

In fact, `legacy` is equivalent to `cipher_compatibility`, except for the case `legacy=0`.

`legacy=0` was introduced to obey the constraint of the _official_ [SQLite Encryption Extension (SEE)](https://sqlite.org/see) that bytes 16 to 23 of the database header are **not** encrypted, because those bytes are accessed by SQLite before any encryption extension gets a chance to decrypt them. The original libraries for the ciphers _SQLCipher_ and _sqleet_ chose to ignore this constraint, although this can lead to invalid database files (even though the chance for this to happen is only _1 : 8192_).

**Reply by little-brother** [reply]
Thanks! 

I thought that versions are different codecs but there are simple pre-defined profiles. So if I call `pragma legacy=2` then all parameters will be switched to [v2-column](https://utelle.github.io/SQLite3MultipleCiphers/docs/ciphers/cipher_sqlcipher/). But if I change `legacy` to `0` then no one parameter will be changed except `legacy` (and `legacy_page_size` will be ignored).

**Reply by utelle** [reply]
> I thought that versions are different codecs but there are simple pre-defined profiles.

Right.

> So if I call `pragma legacy=2` then all parameters will be switched to [v2-column](https://utelle.github.io/SQLite3MultipleCiphers/docs/ciphers/cipher_sqlcipher/).

Yes, exactly. `pragma legacy=2` sets the parameters as required by SQLCipher version 2. If non-default parameters were used for an existing database file created with the original SQLCipher library, they have to be adjusted **after** calling `pragma legacy=2`.

> But if I change `legacy` to `0` then no one parameter will be changed except `legacy` (and `legacy_page_size` will be ignored).

Correct. `pragma legacy=0` just turns legacy mode off and uses the other parameter settings as they are. 

The parameter `legacy_page_size` is required only for legacy SQLCipher database files, because the page size is encoded within the bytes 16 to 23 of the database header. Since the original SQLCipher library encrypts those bytes, the page size value - as read by SQLite - is usually invalid, and therefore SQLite assumes the default page size, which may not be the page size used by SQLCipher.

**Reply by little-brother** [reply]
Thanks, again. The question is closed for me.
